### PR TITLE
Update requirements for SMAC, ConfigSpace, scikit-learn

### LINKED ISSUE #1334: Update SGD regressor loss values for scikit learn 1.0
We need to update the loss name in SGD regressor when switching to scikit learn 1.0.

* [Docs](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html#sklearn-linear-model-sgdregressor)

```python
# Current
loss = CategoricalHyperparameter(
    name="loss",
    choices=[
        "squared_loss", "huber", "epsilon_insensitive", "squared_epsilon_insensitive"
    ],
    default_value="squared_loss",
)

# New
loss = CategoricalHyperparameter(
    name="loss",
    choices=[
        "squared_error", "huber", "epsilon_insensitive", "squared_epsilon_insensitive"
    ],
    default_value="squared_loss",
)
```

> Deprecated since version 1.0: The loss ‘squared_loss’ was deprecated in v1.0 and will be removed in version 1.2. Use loss='squared_error' which is equivalent.

### LINKED ISSUE #1509: Update SMAC, Configspace, scitkit-learn

### PR DESCRIPTION
**Pull request by eddiebergman**
* Closes #1509
* Closes #1334

Issues
---
* The names of some the losses changed, for example with `SGDClassifier`. Changing things to there new values break things for 3.7 as scikit learn does not support this version anymore. I don't want to have a check like `if python_version < (3.8): losses = [...]; else: losses = [...].
* Need to update the metalearning to accomodate this. I have no idea how

---
* Drops Python 3.7
* Lots of test value fixture updates for regressors
* Fixed issue where sparse arrays were not being tested properly to make sure no modification occurs (hash of spares matrix doesn't give values)
* Added test to classifiers to make sure the dont modify data
* Fixed issue where inverse_transform in predict of some models needed to be reshaped.
* Update losses/criterions and api changes according to the two release notes. This was only by looking at `api changes` tag:
    * [1.0.2](https://scikit-learn.org/1.0/whats_new/v1.0.html#changes-1-0)
    * [1.1.1 (latest)](https://scikit-learn.org/stable/whats_new/v1.1.html#changes-1-1)

Notes
---
* Seems there are a few valid losses that are missing as possible hyperparameters for various algorithms
* CV results needs to be updated for sklearn compliance
* Boston dataset soon to be removed in 1.2: https://github.com/scikit-learn/scikit-learn/pull/20729

---
* Need to re-include 3.10 in testing and make releases for it

I think we can just go with this update and then have the next milestone be aiming towards sklearn compliance and usability. Since it's been a while since they've all been reviewed, I think we may have to go do a report on each one and check if there's new possible values and/or new models.

### CODE REVIEW

**Comment by eddiebergman** [comment]
Too out of sync with development, going to start a fresh
